kernel/src/table_properties.rs

```diff
@@ -0,0 +1,373 @@
+//! Delta Table properties. Note this module implements per-table configuration which governs how
+//! table-level capabilities/properties are configured (turned on/off etc.). This is orthogonal to
+//! protocol-level 'table features' which enable or disable reader/writer features (which then
+//! usually must be enabled/configured by table properties).
+//!
+//! For example (from delta's protocol.md): A feature being supported does not imply that it is
+//! active. For example, a table may have the `appendOnly` feature listed in writerFeatures, but it
+//! does not have a table property delta.appendOnly that is set to `true`. In such a case the table
+//! is not append-only, and writers are allowed to change, remove, and rearrange data. However,
+//! writers must know that the table property delta.appendOnly should be checked before writing the
+//! table.
+
+use std::collections::HashMap;
+use std::time::Duration;
+
+use serde::Deserialize;
+
+use crate::expressions::ColumnName;
+use crate::table_features::ColumnMappingMode;
+use crate::DeltaResult;
+
+mod deserialize;
+use deserialize::*;
+
+/// Default num index cols
+pub const DEFAULT_NUM_INDEX_COLS: i32 = 32;
+
+/// Delta table properties. These are parsed from the 'configuration' map in the most recent
+/// 'Metadata' action of a table.
+#[derive(Deserialize, Debug, Clone, Eq, PartialEq, Default)]
+#[serde(default)]
+pub struct TableProperties {
```

Comment:
It seems like we need a way to track unrecognized table properties? If nothing else, the writer needs to propagate them (unchanged) as part of any Metadata action change it produces.

Reply:
adding `unknown_properties: Option<HashMap<String, String>>`